File: frontend/src/services/shop.ts

import api from './api'

export interface ShopInfo {
  id: number
  user_id: number
  name: string
  logo?: string
  address: string
  latitude?: number
  longitude?: number
  business_hours?: string
  notice?: string
  rating: number
  status: number
  monthly_sales: number
  min_order_amount: number
  delivery_fee: number
  delivery_time: string
  discounts: string
  created_at: string
  updated_at: string
}

export interface CategoryInfo {
  id: number
  shop_id: number
  name: string
  sort_order: number
  created_at: string
  products?: ProductInfo[]
}

export interface ProductInfo {
  id: number
  shop_id: number
  category_id?: number
  name: string
  image?: string
  price: number
  original_price?: number
  description?: string
  stock: number
  sales: number
  status: number
  created_at: string
  updated_at: string
}

export interface PageResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

export interface OrderItemInfo {
  id: number
  order_id: number
  product_id: number
  product_name: string
  product_image?: string
  price: number
  quantity: number
}

export interface AddressInfo {
  contact_name: string
  contact_phone: string
  address: string
}

export interface OrderInfo {
  id: number
  order_no: string
  user_id: number
  shop_id: number
  rider_id?: number
  address: string
  latitude?: number
  longitude?: number
  phone: string
  remark?: string
  total_amount: number
  delivery_fee: number
  status: string
  created_at?: string
  updated_at?: string
  shop_name?: string
  shop_image?: string
  items?: OrderItemInfo[]
  address_info?: AddressInfo
}

export interface ShopStats {
  total_orders: number
  total_revenue: number
  pending_orders: number
  rating: number
}

export interface ShopDetail extends ShopInfo {
  categories?: CategoryInfo[]
}

export const shopApi = {
  apply: (data: {
    name: string
    logo?: string
    address: string
    latitude?: number
    longitude?: number
    business_hours?: string
    notice?: string
  }) => api.post<ShopInfo>('/shop/apply', data),

  getMyShop: () => api.get<ShopInfo>('/shop/my'),

  updateMyShop: (data: {
    name?: string
    logo?: string
    address?: string
    latitude?: number
    longitude?: number
    business_hours?: string
    notice?: string
  }) => api.put<ShopInfo>('/shop/my', data),

  listShops: (params?: {
    page?: number
    page_size?: number
    keyword?: string
    status?: number
  }) => api.get<PageResponse<ShopInfo>>('/shop/list', { params }),

  getShopDetail: (shopId: number) => api.get<ShopDetail>(`/shop/${shopId}`),

  createCategory: (data: { shop_id: number; name: string; sort_order?: number }) =>
    api.post<CategoryInfo>('/shop/category', data),

  listCategories: (shopId: number) => api.get<CategoryInfo[]>(`/shop/category/${shopId}`),

  updateCategory: (categoryId: number, data: { name?: string; sort_order?: number }) =>
    api.put<CategoryInfo>(`/shop/category/${categoryId}`, data),

  deleteCategory: (categoryId: number) => api.delete(`/shop/category/${categoryId}`),

  createProduct: (data: {
    shop_id: number
    category_id?: number
    name: string
    image?: string
    price: number
    original_price?: number
    description?: string
    stock: number
  }) => api.post<ProductInfo>('/shop/product', data),

  listProducts: (shopId: number, params?: {
    page?: number
    page_size?: number
    keyword?: string
    category_id?: number
    status?: number
  }) => api.get<PageResponse<ProductInfo>>(`/shop/product/${shopId}`, { params }),

  getProductDetail: (productId: number) => api.get<ProductInfo>(`/shop/product/detail/${productId}`),

  updateProduct: (productId: number, data: {
    category_id?: number
    name?: string
    image?: string
    price?: number
    original_price?: number
    description?: string
    stock?: number
    status?: number
  }) => api.put<ProductInfo>(`/shop/product/${productId}`, data),

  deleteProduct: (productId: number) => api.delete(`/shop/product/${productId}`),

  getShopOrders: (params?: {
    page?: number
    page_size?: number
    status?: string
  }) => api.get<PageResponse<OrderInfo>>('/shop/my/orders', { params }),

  getShopOrderDetail: (orderId: number) => api.get<OrderInfo>(`/shop/my/orders/${orderId}`),

  acceptOrder: (orderId: number) => api.put<OrderInfo>(`/shop/my/orders/${orderId}/accept`),

  rejectOrder: (orderId: number, reason: string) => api.put<OrderInfo>(`/shop/my/orders/${orderId}/reject`, { reason }),

  orderReady: (orderId: number) => api.put<OrderInfo>(`/shop/my/orders/${orderId}/ready`),

  getMyStats: () => api.get<ShopStats>('/shop/my/stats'),

  cancelOrder: (orderId: number) => api.put<OrderInfo>(`/orders/${orderId}/cancel`),
}

export const adminApi = {
  approveShop: (shopId: number) => api.put<ShopInfo>(`/admin/shop/${shopId}/approve`),
  rejectShop: (shopId: number) => api.put<ShopInfo>(`/admin/shop/${shopId}/reject`),
  listPendingShops: (params?: { page?: number; page_size?: number; keyword?: string }) =>
    api.get<PageResponse<ShopInfo>>('/admin/shop/pending', { params }),
  listAdminOrders: (params?: { page?: number; page_size?: number; status?: string }) =>
    api.get<PageResponse<OrderInfo>>('/admin/orders', { params }),
  getAdminOrderDetail: (orderId: number) => api.get<OrderInfo>(`/admin/orders/${orderId}`),
}

File: backend/app/schemas/shop.py

from decimal import Decimal
from pydantic import Field, field_validator
from typing import Optional, List
from datetime import datetime
from app.schemas.base import BaseSchema, DecimalField
from app.utils.sanitizer import strip_all_tags, sanitize_limited_html, strip_dangerous_content


class ShopCreate(BaseSchema):
    name: str = Field(..., min_length=2, max_length=100)
    logo: Optional[str] = None
    address: str = Field(..., min_length=5, max_length=255)
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    business_hours: Optional[str] = None
    business_days: Optional[str] = None
    notice: Optional[str] = None

    @field_validator("name")
    @classmethod
    def sanitize_name(cls, v: str) -> str:
        """Strip dangerous HTML from shop name (SEC-REFORM-07)."""
        return strip_dangerous_content(v)

    @field_validator("notice")
    @classmethod
    def sanitize_notice(cls, v: Optional[str]) -> Optional[str]:
        """Allow limited safe HTML in shop notice (SEC-REFORM-07)."""
        if v is not None:
            return sanitize_limited_html(v)
        return v


class ShopUpdate(BaseSchema):
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    logo: Optional[str] = None
    address: Optional[str] = Field(None, min_length=5, max_length=255)
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    business_hours: Optional[str] = None
    business_days: Optional[str] = None
    notice: Optional[str] = None
    min_order_amount: Optional[DecimalField] = None
    delivery_fee: Optional[DecimalField] = None
    delivery_time: Optional[str] = None
    discounts: Optional[str] = None

    @field_validator("name")
    @classmethod
    def sanitize_name(cls, v: Optional[str]) -> Optional[str]:
        """Strip dangerous HTML from shop name (SEC-REFORM-07)."""
        if v is not None:
            return strip_dangerous_content(v)
        return v

    @field_validator("notice")
    @classmethod
    def sanitize_notice(cls, v: Optional[str]) -> Optional[str]:
        """Allow limited safe HTML in shop notice (SEC-REFORM-07)."""
        if v is not None:
            return sanitize_limited_html(v)
        return v


class ShopInfo(BaseSchema):
    id: int
    user_id: int
    name: str
    logo: Optional[str] = None
    address: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    business_hours: Optional[str] = None
    business_days: Optional[str] = None
    notice: Optional[str] = None
    rating: float
    status: int
    monthly_sales: int = 0
    min_order_amount: DecimalField = Decimal("20.00")
    delivery_fee: DecimalField = Decimal("3.00")
    delivery_time: str = "30分钟"
    discounts: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class ShopDetail(BaseSchema):
    id: int
    user_id: int
    name: str
    logo: Optional[str] = None
    address: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    business_hours: Optional[str] = None
    business_days: Optional[str] = None
    notice: Optional[str] = None
    rating: float
    status: int
    monthly_sales: int = 0
    min_order_amount: DecimalField = Decimal("20.00")
    delivery_fee: DecimalField = Decimal("3.00")
    delivery_time: str = "30分钟"
    discounts: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    categories: Optional[List["CategoryInfo"]] = None


class CategoryCreate(BaseSchema):
    shop_id: int
    name: str = Field(..., min_length=1, max_length=50)
    sort_order: Optional[int] = 0


class CategoryUpdate(BaseSchema):
    name: Optional[str] = Field(None, min_length=1, max_length=50)
    sort_order: Optional[int] = None


class CategoryInfo(BaseSchema):
    id: int
    shop_id: int
    name: str
    sort_order: int
    created_at: datetime
    products: Optional[List["ProductInfo"]] = None


class ProductCreate(BaseSchema):
    shop_id: int
    category_id: Optional[int] = None
    name: str = Field(..., min_length=1, max_length=100)
    image: Optional[str] = None
    price: DecimalField = Field(..., ge=0)
    original_price: Optional[DecimalField] = None
    description: Optional[str] = None
    stock: int = Field(0, ge=0)

    @field_validator("name")
    @classmethod
    def sanitize_name(cls, v: str) -> str:
        """Strip dangerous HTML from product name (SEC-REFORM-07)."""
        return strip_dangerous_content(v)

    @field_validator("description")
    @classmethod
    def sanitize_description(cls, v: Optional[str]) -> Optional[str]:
        """Allow limited safe HTML in product description (SEC-REFORM-07)."""
        if v is not None:
            return sanitize_limited_html(v)
        return v


class ProductUpdate(BaseSchema):
    category_id: Optional[int] = None
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    image: Optional[str] = None
    price: Optional[DecimalField] = Field(None, ge=0)
    original_price: Optional[DecimalField] = None
    description: Optional[str] = None
    stock: Optional[int] = Field(None, ge=0)
    status: Optional[int] = None

    @field_validator("name")
    @classmethod
    def sanitize_name(cls, v: Optional[str]) -> Optional[str]:
        """Strip dangerous HTML from product name (SEC-REFORM-07)."""
        if v is not None:
            return strip_dangerous_content(v)
        return v

    @field_validator("description")
    @classmethod
    def sanitize_description(cls, v: Optional[str]) -> Optional[str]:
        """Allow limited safe HTML in product description (SEC-REFORM-07)."""
        if v is not None:
            return sanitize_limited_html(v)
        return v


class ProductInfo(BaseSchema):
    id: int
    shop_id: int
    category_id: Optional[int] = None
    name: str
    image: Optional[str] = None
    price: DecimalField
    original_price: Optional[DecimalField] = None
    description: Optional[str] = None
    stock: int
    sales: int
    tags: Optional[str] = None
    rating: float = 0.0
    status: int
    created_at: datetime
    updated_at: datetime


class ShopListQuery(BaseSchema):
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    keyword: Optional[str] = None
    status: Optional[int] = None


class ProductListQuery(BaseSchema):
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    keyword: Optional[str] = None
    category_id: Optional[int] = None
    status: Optional[int] = None
    min_price: Optional[float] = None
    max_price: Optional[float] = None
    sort_by: Optional[str] = None  # price, sales, rating


class ShopReview(BaseSchema):
    pass
